design-documents/testing-strategy.md

```diff
@@ -0,0 +1,39 @@
+## Updated Magento Commerce Testing Strategy
+
+### Terms
+
+<!-- Describe any new terms used in the document -->
+
+### Overview
+
+<!-- Describe what we are trying to solve and a few sentence summary of design approach. -->
+
+Despite running 20,000 - 30,000 tests per pull request, we have little to no confidence in the stability of the product. This means that we still need a manual regression process that lasts weeks in the best of cases. 
+
+A sizable portion of this can be attributed to our current high-level testing mentality that prioritizes system-level black box testing over other types of tests. While these types of tests are necessary and good to have, they only provide implicit coverage to the code paths of the concrete implementations that happen to be loaded in the high-level black box tests (they are also expensive/difficult to maintain). We still need explicit coverage for the concrete implementations that extension developers have hard dependencies on. This is true for the simple code paths as well as variations both positive and negative flows.
+
+### Design
+
+#### New Testing Strategy:
+
+The new strategy for testing relies on having test coverage at each layer. The idea is to have many small and fast tests. Therefore, tests MUST be written in the following order of granularity and the higher-level the test, the fewer tests there will be:
+
+1. Code MUST be tested on its own. This MAY be done through unit or integration testing as appropriate. This level of test coverage ensures that the code can properly function on its own.  
+1. Code MUST be shown to work with known collaborators. This MUST be done through integration tests and ideally SHOULD involve as few components as possible per test case. This level of tests ensures that the code will work with its known associates. Examples of such associates would include aggregators, composites, etc.
```

Comment:
I would suggest to add more clarity on how using integration test here would be different than in previous item if you think that it's important. If following previous item sometimes would mean following this one as well, than it might be ok.

Reply:
I'm not sure I understand this question. The idea here is the verifying that code works in the ways that it could be used. First it could be used on its own (such as a di injected hard dependency) so we need to make sure it behaves correctly. Secondly, it could be used with something else (such as with a pool, aggregator, composite, renderer, chain, etc) and it may not interact well with other components in this way. This is the more traditional idea of an integration test.